### PR TITLE
[WIP][Do no review] Tools: Testbech: Add IPC4 support

### PR DESCRIPTION
**Pull request by singalsu**
First somewhat cleaned up version that runs, but needs still more work. Usage e.g.

```
cd sof/tools/test/audio
../../testbench/build_testbench/install/bin/testbench -r 48000 -R 48000 -c 2 -n 2 -b S32_LE \
-t ../../build_tools/topology/topology2/development/sof-hda-benchmark-dcblock32.tplg \
-p 1,2 -i i1.raw,i1.raw,i2.raw,i3.raw,i4.raw -o o1.raw,o2.raw,o3.raw,o4.raw,o5.raw 

```
To-do:
- Bytes control is not passed
- Only one processing component is possible, now hard-coded to dcblock, need kind of rimage uuid to index to uuid and fw uuid based component initialize mechanism
- Need still to solve IPC3 testbench build
- Simplify command line interface to run a simple playback test, no "-p 1,2 -i i1.raw,i1.raw,i2.raw,i3.raw,i4.raw -o o1.raw,o2.raw,o3.raw,o4.raw,o5.raw" but just "-i in.raw -o out.raw".
- Fix memory leaks (valgrind)
- Fix code branch based on uninitialized (valgrind)
- Trace on/off (broken in main too)

### CODE REVIEW

File: src/arch/host/CMakeLists.txt [hunk 1]
```diff
@@ -18,7 +18,7 @@ if (supports_implicit_fallthrough)
 endif()
 
 # C & ASM flags
-target_compile_options(sof_options INTERFACE -g -O3 -fPIC -DPIC -std=c99 -std=gnu99 -fgnu89-inline
+target_compile_options(sof_options INTERFACE -g -fPIC -DPIC -std=c99 -std=gnu99 -fgnu89-inline
```
Comment (lgirdwood) [inline src/arch/host/CMakeLists.txt:21]: long term we can inherit the optimization level from Kconfig

File: src/ipc/ipc4/handler.c [hunk 2]
```diff
@@ -534,6 +534,9 @@ static void ipc_compound_msg_done(uint32_t msg_id, int error)
 
 static int ipc_wait_for_compound_msg(void)
 {
+#if  CONFIG_LIBRARY
+	atomic_set(&msg_data.delayed_reply, 0);
```
Comment (lgirdwood) [inline src/ipc/ipc4/handler.c:538]: Btw, where do we clear this in DSP mode ? i.e. does this need to be executed on all targets ?

File: src/audio/pipeline/pipeline-stream.c [hunk 3]
```diff
@@ -486,6 +486,7 @@ static int pipeline_comp_trigger(struct comp_dev *current,
 			 * Initialization delay is only used with SSP, where we
 			 * don't use more than one DAI per copier
 			 */
+#if !CONFIG_LIBRARY
```
Comment (lgirdwood) [inline src/audio/pipeline/pipeline-stream.c:489]: Can we provide this API in the file module instead, even if it returns 0 today.

File: tools/testbench/topology.c [hunk 4]
```diff
@@ -369,23 +379,33 @@ static int tb_register_fileread(struct testbench_prm *tp,
...
+	file_uuid->b = 0x75aa;
+	file_uuid->c = 0x4ce8;
+	file_uuid->d[0] = 0x9d;
+	file_uuid->d[1] = 0xbe;
+	file_uuid->d[2] = 0xd8;
+	file_uuid->d[3] = 0xda;
+	file_uuid->d[4] = 0x08;
+	file_uuid->d[5] = 0xa6;
+	file_uuid->d[6] = 0x98;
+	file_uuid->d[7] = 0xc2;
```
Comment (lgirdwood) [inline tools/testbench/topology.c]: I think there is a macro that will pack this for you now. @kv2019i @lyakh ?

**Review by lgirdwood — COMMENTED** [review]
Ok, looks like most of the small fix patches are merge ready (you may want to put them in another PR), whilst some rebase/squash is needed on others.
@ranj063 pls review

**Comment by singalsu** [comment]
> Ok, looks like most of the small fix patches are merge ready (you may want to put them in another PR), whilst some rebase/squash is needed on others. @ranj063 pls review

I will next try to separate this IPC4 related work from existing IPC3 testbench that would remain mostly unchanged except file component that changes to module adapter. So, there would be two executables: testbench and testbench-ipc4. When I started this I was thinking IPC4/3 choice would be run-time but such is not possible due to other SOF parts.

File: src/audio/pipeline/pipeline-stream.c (hunk 3 above)
Comment (singalsu) [reply src/audio/pipeline/pipeline-stream.c:489]: Yep, I started to separate file component module adapter changes into a separate PR. It was quite simple to get dai_get_init_delay_ms() to pass without duplicating much of DAI data structures.

**Comment by singalsu** [comment]
> @singalsu see also this for Cmake fixes for testbench #8892

Thanks I didn't know -- I wonder if it could help to build with some scripts/rebuild-testbech.sh a desired IPC version. Or somehow build both to different executable paths. There's now new testbench PR version in: #9483. I will close this one later.

**Pull request closed without merging**